os_stub/cryptlib_mbedtls/pk/x509.c

```diff
@@ -1576,6 +1593,120 @@ int32_t libspdm_x509_compare_date_time(const void *date_time1, const void *date_
+        { ADD_STRLEN( "L" ), ADD_STRLEN( MBEDTLS_OID_AT_LOCALITY ), MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "R" ), ADD_STRLEN( MBEDTLS_OID_PKCS9_EMAIL ), MBEDTLS_ASN1_IA5_STRING },
+        { ADD_STRLEN( "OU" ), ADD_STRLEN( MBEDTLS_OID_AT_ORG_UNIT ), MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "ST" ), ADD_STRLEN( MBEDTLS_OID_AT_STATE ), MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "emailAddress" ), ADD_STRLEN( MBEDTLS_OID_PKCS9_EMAIL ),
+          MBEDTLS_ASN1_IA5_STRING },
+        { ADD_STRLEN( "serialNumber" ), ADD_STRLEN( MBEDTLS_OID_AT_SERIAL_NUMBER ),
+          MBEDTLS_ASN1_PRINTABLE_STRING },
+        { ADD_STRLEN( "postalAddress" ), ADD_STRLEN( MBEDTLS_OID_AT_POSTAL_ADDRESS ),
+          MBEDTLS_ASN1_PRINTABLE_STRING },
+        { ADD_STRLEN( "postalCode" ), ADD_STRLEN( MBEDTLS_OID_AT_POSTAL_CODE ),
+          MBEDTLS_ASN1_PRINTABLE_STRING },
+        { ADD_STRLEN( "dnQualifier" ), ADD_STRLEN( MBEDTLS_OID_AT_DN_QUALIFIER ),
+          MBEDTLS_ASN1_PRINTABLE_STRING },
+        { ADD_STRLEN( "title" ), ADD_STRLEN( MBEDTLS_OID_AT_TITLE ), MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "SN" ), ADD_STRLEN( MBEDTLS_OID_AT_SUR_NAME ), MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "GN" ), ADD_STRLEN( MBEDTLS_OID_AT_GIVEN_NAME ),  MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "initials" ), ADD_STRLEN( MBEDTLS_OID_AT_INITIALS ),
+          MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "pseudonym" ), ADD_STRLEN( MBEDTLS_OID_AT_PSEUDONYM ),
+          MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "generationQualifier" ), ADD_STRLEN( MBEDTLS_OID_AT_GENERATION_QUALIFIER ),
+          MBEDTLS_ASN1_UTF8_STRING },
+        { ADD_STRLEN( "DC" ), ADD_STRLEN( MBEDTLS_OID_DOMAIN_COMPONENT ),
+          MBEDTLS_ASN1_IA5_STRING },
+        { NULL, 0, NULL, 0, MBEDTLS_ASN1_NULL }
+    };
+
+    end = ptr + obj_len;
+
+    while(ptr != end) {
+        /*SET*/
+        ret = libspdm_asn1_get_tag(&ptr, end, &obj_len,
+                                   LIBSPDM_CRYPTO_ASN1_SET | LIBSPDM_CRYPTO_ASN1_CONSTRUCTED);
+        if (!ret) {
+            return false;
+        }
+        internal_p = ptr;
+        /*move to next SET*/
+        ptr += obj_len;
+
+        /*sequece*/
+        ret = libspdm_asn1_get_tag(&internal_p, end, &obj_len,
+                                   LIBSPDM_CRYPTO_ASN1_SEQUENCE | LIBSPDM_CRYPTO_ASN1_CONSTRUCTED);
+        if (!ret) {
+            return false;
+        }
+
+        /*OID*/
+        ret = libspdm_asn1_get_tag(&internal_p, end, &obj_len, LIBSPDM_CRYPTO_ASN1_OID);
+        if (!ret) {
+            return false;
+        }
+
+        for (cur = x509_attrs; cur->name != NULL; cur++) {
+            if ((cur->oid_len == obj_len) &&
+                (libspdm_consttime_is_mem_equal(cur->oid, internal_p, obj_len))) {
+                /*Concat subject string*/
+
+                /*for example: CN=*/
+                libspdm_copy_mem(buffer, buff_len, cur->name, cur->name_len);
```

Comment:
What if buff_len is big enough ?

Reply:
There is a buffer_len subtract operation for every copy.
```
                buff_len = (int32_t)(buff_len - cur->name_len);
                buffer += cur->name_len;
                *buffer = '=';
                buff_len--;
                buffer++;
```

And there is a check for buffer_len:
```
                if (buff_len < 0) {
                    LIBSPDM_DEBUG((LIBSPDM_DEBUG_INFO,"the buffer is too small"));
                    return false;
                }
```